contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,139 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+// EpochTokenLocker saveguards tokens for applications with constant-balances during discrete epochs
+// It allows to deposit token which become credited in the next epoch and allows to request a token-withdraw
+// which becomes claimable after the current epoch expired.
+contract EpochTokenLocker {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event WithdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits; // deposits will be credited in any next epoch, i.e. currentStateIndex > stateIndex
+        PendingFlux pendingWithdraws; // withdraws are allowed in any next epoch , i.e. currentStateIndex > stateIndex
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount = balanceStates[msg.sender][token].pendingDeposits.amount
+            .add(amount);
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function requestWithdraw(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
```

Comment:
Are we using events here in the same way we were for the Snapp contracts? Is this purely for the convenience of unit testing? I guess my question is, are these events intended to be used a reference for a future transaction?